src/util/check.h

```diff
@@ -46,4 +46,14 @@ class NonFatalCheckError : public std::runtime_error
 #error "Cannot compile without assertions!"
 #endif
 
+/** Helper for Assert(). TODO remove in C++14 and replace `decltype(get_pure_r_value(val))` with `T` (templated lambda) */
+template <typename T>
+T get_pure_r_value(T&& val)
+{
+    return std::forward<T>(val);
+}
+
+/** Identity function. Abort if the value compares equal to zero */
+#define Assert(val) [&]() -> decltype(get_pure_r_value(val))& { auto& check = (val); assert(#val && check); return check; }()
```

Comment:
In commit "util: Add Assert identity function" (fa6ef701adba1cb48535cac25fd43c742a82e40d)

Note: `decltype(get_pure_r_value(val))` is needed here in case the `val` expression is const, because just `decltype(val)` drops the const and results in an error trying to return a const reference from a lambda effectively declared to return a non-const reference. I think returning decltype(auto) would be another way to work around this once we update past c++11

Reply:
Thanks for the hint. Whoever plays code-golf after we switch to C++17 should consider this.